File: src/test/java/com/github/jactor/persistence/service/BlogServiceTest.java
package com.github.jactor.persistence.service;

import static com.github.jactor.persistence.entity.BlogEntity.aBlog;
import static com.github.jactor.persistence.entity.BlogEntryEntity.aBlogEntry;
import static java.time.LocalDate.now;
import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.Assertions.assertAll;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import com.github.jactor.persistence.dto.BlogDto;
import com.github.jactor.persistence.dto.BlogEntryDto;
import com.github.jactor.persistence.dto.PersistentDto;
import com.github.jactor.persistence.dto.UserInternalDto;
import com.github.jactor.persistence.entity.BlogEntity;
import com.github.jactor.persistence.entity.BlogEntryEntity;
import com.github.jactor.persistence.repository.BlogEntryRepository;
import com.github.jactor.persistence.repository.BlogRepository;
import java.util.Collections;
import java.util.List;
import java.util.Optional;
import java.util.function.Supplier;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.ArgumentCaptor;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

@ExtendWith(MockitoExtension.class)
@DisplayName("A BlogService")
class BlogServiceTest {

  @InjectMocks
  private BlogService blogServiceToTest;
  @Mock
  private BlogRepository blogRepositoryMock;
  @Mock
  private BlogEntryRepository blogEntryRepositoryMock;
  @SuppressWarnings("unused") // used by mockito
  @Mock
  private UserService userServiceMock;

  @Test
  @DisplayName("should map blog to dto")
  void shouldMapBlogToDto() {
    Optional<BlogEntity> blogEntity = Optional.of(aBlog(new BlogDto(new PersistentDto(), null, "full speed ahead", null)));
    when(blogRepositoryMock.findById(1001L)).thenReturn(blogEntity);

    BlogDto blog = blogServiceToTest.find(1001L).orElseThrow(mockError());

    assertThat(blog.getTitle()).as("title").isEqualTo("full speed ahead");
  }

  @Test
  @DisplayName("should map blog entry to dto")
  void shouldMapFoundBlogToDto() {
    BlogEntryDto blogEntryDto = new BlogEntryDto(new PersistentDto(), new BlogDto(), "me", "too");
    Optional<BlogEntryEntity> anEntry = Optional.of(aBlogEntry(blogEntryDto));
    when(blogEntryRepositoryMock.findById(1001L)).thenReturn(anEntry);

    BlogEntryDto blogEntry = blogServiceToTest.findEntryBy(1001L).orElseThrow(mockError());

    assertAll(
        () -> assertThat(blogEntry.getCreatorName()).as("creator name").isEqualTo("me"),
        () -> assertThat(blogEntry.getEntry()).as("entry").isEqualTo("too")
    );
  }

  private Supplier<AssertionError> mockError() {
    return () -> new AssertionError("missed mocking?");
  }

  @Test
  @DisplayName("should find blogs for title")
  void shouldFindBlogsForTitle() {
    List<BlogEntity> blogsToFind = Collections.singletonList(aBlog(new BlogDto(new PersistentDto(), null, "Star Wars", null)));
    when(blogRepositoryMock.findBlogsByTitle("Star Wars")).thenReturn(blogsToFind);

    List<BlogDto> blogForTitle = blogServiceToTest.findBlogsBy("Star Wars");

    assertThat(blogForTitle).hasSize(1);
  }

  @Test
  @DisplayName("should map blog entries to a list of dto")
  void shouldMapBlogEntriesToListOfDto() {
    List<BlogEntryEntity> blogEntryEntities = Collections.singletonList(
        aBlogEntry(new BlogEntryDto(new PersistentDto(), new BlogDto(), "you", "too"))
    );

    when(blogEntryRepositoryMock.findByBlog_Id(1001L)).thenReturn(blogEntryEntities);

    List<BlogEntryDto> blogEntries = blogServiceToTest.findEntriesForBlog(1001L);

    assertAll(
        () -> assertThat(blogEntries).as("entries").hasSize(1),
        () -> assertThat(blogEntries.get(0).getCreatorName()).as("creator name").isEqualTo("you"),
        () -> assertThat(blogEntries.get(0).getEntry()).as("entry").isEqualTo("too")
    );
  }

  @Test
  @DisplayName("should save BlogDto as BlogEntity")
  void shouldSaveBlogDtoAsBlogEntity() {
    BlogDto blogDto = new BlogDto();
    blogDto.setCreated(now());
    blogDto.setTitle("some blog");
    blogDto.setUserInternal(new UserInternalDto());

    blogServiceToTest.saveOrUpdate(blogDto);

    ArgumentCaptor<BlogEntity> argCaptor = ArgumentCaptor.forClass(BlogEntity.class);
    verify(blogRepositoryMock).save(argCaptor.capture());
    BlogEntity blogEntity = argCaptor.getValue();

    assertAll(
        () -> assertThat(blogEntity.getCreated()).as("created").isEqualTo(now()),
        () -> assertThat(blogEntity.getTitle()).as("title").isEqualTo("some blog"),
        () -> assertThat(blogEntity.getUser()).as("user").isNotNull()
    );
  }

  @Test
  @DisplayName("should save BlogEntryDto as BlogEntryEntity")
  void shouldSaveBlogEntryDtoAsBlogEntryEntity() {
    BlogEntryDto blogEntryDto = new BlogEntryDto();
    blogEntryDto.setBlog(new BlogDto());
    blogEntryDto.setCreatorName("me");
    blogEntryDto.setEntry("if i where a rich man...");

    blogServiceToTest.saveOrUpdate(blogEntryDto);

    ArgumentCaptor<BlogEntryEntity> argCaptor = ArgumentCaptor.forClass(BlogEntryEntity.class);
    verify(blogEntryRepositoryMock).save(argCaptor.capture());
    BlogEntryEntity blogEntryEntity = argCaptor.getValue();

    assertAll(
        () -> assertThat(blogEntryEntity.getBlog()).as("blog").isNotNull(),
        () -> assertThat(blogEntryEntity.getCreatorName()).as("creator name").isEqualTo("me"),
        () -> assertThat(blogEntryEntity.getEntry()).as("entry").isEqualTo("if i where a rich man...")
    );
  }
}

File: persistence/src/test/kotlin/com/github/jactor/persistence/service/BlogServiceTest.kt
package com.github.jactor.persistence.service

import java.time.LocalDate
import java.util.Optional
import java.util.UUID
import org.junit.jupiter.api.Test
import com.github.jactor.persistence.dto.BlogModel
import com.github.jactor.persistence.dto.BlogEntryModel
import com.github.jactor.persistence.dto.PersistentModel
import com.github.jactor.persistence.dto.UserModel
import com.github.jactor.persistence.entity.BlogBuilder
import com.github.jactor.persistence.entity.BlogEntity
import com.github.jactor.persistence.entity.BlogEntryEntity
import com.github.jactor.persistence.repository.BlogEntryRepository
import com.github.jactor.persistence.repository.BlogRepository
import assertk.assertAll
import assertk.assertThat
import assertk.assertions.hasSize
import assertk.assertions.isEqualTo
import assertk.assertions.isNotNull
import assertk.assertions.isNull
import io.mockk.every
import io.mockk.mockk
import io.mockk.slot

internal class BlogServiceTest {

    private val blogEntryRepositoryMockk: BlogEntryRepository = mockk {}
    private val blogRepositoryMockk: BlogRepository = mockk {}
    private val userServiceMockk: UserService = mockk {}

    private val blogServiceToTest: BlogService = DefaultBlogService(
        blogRepository = blogRepositoryMockk,
        blogEntryRepository = blogEntryRepositoryMockk,
        userService = userServiceMockk
    )

    private val uuid: UUID = UUID.randomUUID()


    @Test
    fun `should map blog to dto`() {
        val blogEntity = BlogBuilder.new(
            blogModel = BlogModel(PersistentModel(), null, "full speed ahead", null)
        ).buildBlogEntity()

        every { blogRepositoryMockk.findById(uuid) } returns Optional.of(blogEntity)

        val (_, _, title) = blogServiceToTest.find(uuid) ?: throw AssertionError("missed mocking?")

        assertThat(title).isEqualTo("full speed ahead")
    }

    @Test
    fun `should map blog entry to dto`() {
        val blogEntryModel = BlogEntryModel(PersistentModel(), BlogModel(), "me", "too")
        val anEntry = BlogBuilder.new().withEntry(blogEntryModel = blogEntryModel).buildBlogEntryEntity()

        every { blogEntryRepositoryMockk.findById(uuid) } returns Optional.of(anEntry)

        val (_, _, creatorName, entry) = blogServiceToTest.findEntryBy(uuid)
            ?: throw AssertionError("missed mocking?")

        assertAll {
            assertThat(creatorName).isEqualTo("me")
            assertThat(entry).isEqualTo("too")
        }
    }

    @Test
    fun `should find blogs for title`() {
        val blogsToFind = listOf(BlogBuilder.new(blogModel = BlogModel(title = "Star Wars")).buildBlogEntity())

        every { blogRepositoryMockk.findBlogsByTitle("Star Wars") } returns blogsToFind

        val blogForTitle = blogServiceToTest.findBlogsBy("Star Wars")

        assertThat(blogForTitle).hasSize(1)
    }

    @Test
    fun `should map blog entries to a list of dto`() {
        val blogEntryEntities: List<BlogEntryEntity?> = listOf(
            BlogBuilder.new()
                .withEntry(blogEntryModel = BlogEntryModel(PersistentModel(), BlogModel(), "you", "too"))
                .buildBlogEntryEntity()
        )

        every { blogEntryRepositoryMockk.findByBlog_Id(uuid) } returns blogEntryEntities

        val blogEntries = blogServiceToTest.findEntriesForBlog(uuid)

        assertAll {
            assertThat(blogEntries).hasSize(1)
            assertThat(blogEntries[0].creatorName).isEqualTo("you")
            assertThat(blogEntries[0].entry).isEqualTo("too")
        }
    }

    @Test
    fun `should save BlogDto as BlogEntity`() {
        val blogEntitySlot = slot<BlogEntity>()
        val blogModel = BlogModel()

        blogModel.created = LocalDate.now()
        blogModel.title = "some blog"
        blogModel.userInternal = UserModel(username = "itsme")

        every { userServiceMockk.find(username = any()) } returns null
        every { blogRepositoryMockk.save(capture(blogEntitySlot)) } returns BlogEntity(blogModel)

        blogServiceToTest.saveOrUpdate(blogModel = blogModel)
        val blogEntity = blogEntitySlot.captured

        assertAll {
            assertThat(blogEntity.created).isEqualTo(LocalDate.now())
            assertThat(blogEntity.title).isEqualTo("some blog")
            assertThat(blogEntity.user).isNull()
        }
    }

    @Test
    fun `should save BlogEntryDto as BlogEntryEntity`() {
        val blogEntryEntitySlot = slot<BlogEntryEntity>()
        val blogEntryModel = BlogEntryModel()
        blogEntryModel.blog = BlogModel(
            persistent = PersistentMoDto(id = UUID.randomUUID()),
            userInternal = UserModel(username = "itsme")
        )

        blogEntryDto.creatorName = "me"
        blogEntryDto.entry = "if i where a rich man..."

        every { userServiceMockk.find(username = any()) } returns null
        every { blogEntryRepositoryMockk.save(capture(blogEntryEntitySlot)) } returns BlogEntryEntity(blogEntryModel)

        blogServiceToTest.saveOrUpdate(blogEntryModel)
        val blogEntryEntity = blogEntryEntitySlot.captured

        assertAll {
            assertThat(blogEntryEntity.blog).isNotNull()
            assertThat(blogEntryEntity.creatorName).isEqualTo("me")
            assertThat(blogEntryEntity.entry).isEqualTo("if i where a rich man...")
        }
    }
}
